src/atomic.rs

```diff
@@ -1,160 +1,380 @@
+            Ok(())
+        })?;
+
+        Ok(input_id)
+    }
+
+    pub fn cancel_atomic_exec<'a, S>(
+        &mut self,
+        bs: &impl Blockstore,
+        input_id: AtomicInputID,
+        input_fn: impl FnOnce(AtomicInput) -> Box<dyn Iterator<Item = &'a mut S>>,
+    ) -> anyhow::Result<()>
+    where
+        S: LockableState + 'a,
+    {
+        // Consume own input ID and retrieve the associated data
+        let AtomicInputEntry { input, .. } = self.input_ids.modify(bs, |m| {
+            let k = BytesKey::from(input_id.bytes());
+            let (_, v) = m
+                .delete(&k)?
+                .ok_or_else(|| anyhow::anyhow!("unexpected input ID"))?;
+            Ok(v)
+        })?;
+
+        // Get the state and ensure it's unlocked
+        let state_iter = input_fn(input);
+        state_iter.for_each(|s| {
+            if s.is_locked() {
+                s.unlock().unwrap();
+            }
+        });
+
+        Ok(())
+    }
+
+    pub fn prepare_atomic_exec<'a, S, I>(
+        &mut self,
+        bs: &impl Blockstore,
+        own_input_id: AtomicInputID,
+        input_ids: &HashMap<IPCAddress, AtomicInputID>,
+        input_fn: impl FnOnce(AtomicInput) -> (I, Box<dyn Iterator<Item = &'a mut S>>),
+        output_fn: impl FnOnce(I) -> anyhow::Result<AtomicOutput>,
+    ) -> anyhow::Result<AtomicExecID>
+    where
+        S: 'a + LockableState,
+    {
+        // Consume own input ID and retrieve the associated data
+        let AtomicInputEntry {
+            unlocked_state_cids,
+            input,
+        } = self.input_ids.modify(bs, |m| {
+            let k = BytesKey::from(own_input_id.bytes());
+            let (_, v) = m
+                .delete(&k)?
+                .ok_or_else(|| anyhow::anyhow!("unexpected own input ID"))?;
+            Ok(v)
+        })?;
+
+        // Get the input and the state; check that the state has not
+        // changed and ensure it is locked
+        let (input, state_iter) = input_fn(input);
+        let unlocked_state_cid_iter = state_iter.filter(|s| !s.is_locked()).map(|s| {
+            let cid = s.cid();
+            s.lock().unwrap();
+            cid
+        });
+        if !unlocked_state_cid_iter.eq(unlocked_state_cids) {
+            anyhow::bail!("state CID mismatch");
+        }
+
+        // Compute the atomic execution ID; produce and store the
+        // output
+        let exec_id = Self::compute_exec_id(input_ids);
+        self.exec_ids.modify(bs, |m| {
+            let k = BytesKey::from(exec_id.bytes());
+            let output = output_fn(input)?;
+            let v = m.set(k, AtomicOutputEntry { output })?;
+            assert!(v.is_none(), "exec ID collision");
+            Ok(())
+        })?;
+
+        Ok(exec_id)
+    }
+
+    pub fn commit_atomic_exec<'a, S, O, R>(
+        &mut self,
+        bs: &impl Blockstore,
+        exec_id: AtomicExecID,
+        output_fn: impl FnOnce(AtomicOutput) -> (O, Box<dyn Iterator<Item = &'a mut S>>),
```

Comment:
Same as above. We need better descriptions of what these functions are. To double-check my understanding: 
- Output function computes the output for the messages received as input in the atomic execution. 
- Apply functions takes the output and commits it into the state.

Reply:
This `output_fn` only interprets the serialized (raw) representation of the output produced by the other closure supplied to `prepare_atomic_exec`. Check the added comments which may help to see this better.